test/widgets/profile_test.dart

```diff
@@ -0,0 +1,269 @@
+            urlPattern: 'http://example/%(username)s')},
+      );
+
+      await tester.tap(find.text('externalValue'));
+      final expectedMode = defaultTargetPlatform == TargetPlatform.android ?
+        LaunchMode.externalApplication : LaunchMode.platformDefault;
+      check(testBinding.takeLaunchUrlCalls())
+        .single.equals((url: Uri.parse('http://example/externalValue'), mode: expectedMode));
+    });
+
+    testWidgets('page builds; user links to profile', (WidgetTester tester) async {
+      final users = [
+        eg.user(userId: 1, profileData: {
+          0: ProfileFieldUserData(value: '[2]'),
+        }),
+        eg.user(userId: 2, fullName: 'test user'),
+      ];
+      final pushedRoutes = <Route<dynamic>>[];
+      final testNavObserver = TestNavigatorObserver()
+        ..onPushed = (route, prevRoute) => pushedRoutes.add(route);
+
+      await setupPage(tester,
+        users: users,
+        pageUserId: 1,
+        customProfileFields: [mkCustomProfileField(0, CustomProfileFieldType.user)],
+        navigatorObserver: testNavObserver,
+      );
+
+      final textFinder = find.text('test user');
+      check(textFinder.evaluate()).length.equals(1);
+      final fieldContainer = find.ancestor(of: textFinder, matching: find.byType(Column)).first;
+      final targetWidget = find.descendant(of: fieldContainer, matching:find.byType(Avatar));
+      await tester.tap(targetWidget, warnIfMissed: false);
+      check(pushedRoutes).last.isA<WidgetRoute>().page.isA<ProfilePage>().userId.equals(2);
+    });
+
+    testWidgets('page builds; dm links to correct narrow', (WidgetTester tester) async {
+      final pushedRoutes = <Route<dynamic>>[];
+      final testNavObserver = TestNavigatorObserver()
+        ..onPushed = (route, prevRoute) => pushedRoutes.add(route);
+      final expectedNarrow = DmNarrow(
+        allRecipientIds: [eg.selfUser.userId, 1]..sort(),
+        selfUserId: eg.selfUser.userId);
+
+      await setupPage(tester,
+        users: [eg.user(userId: 1)],
+        pageUserId: 1,
+        navigatorObserver: testNavObserver,
+      );
+
+      final targetWidget = find.byIcon(Icons.email);
+      await tester.tap(targetWidget, warnIfMissed: false);
+      check(pushedRoutes).last.isA<WidgetRoute>().page
+        .isA<MessageListPage>()
+        .narrow.equals(expectedNarrow);
+    });
+
+    testWidgets('page builds; user links render multiple avatars', (WidgetTester tester) async {
+      final users = [
+        eg.user(userId: 1, profileData: {
+          0: ProfileFieldUserData(value: '[2,3]'),
+        }),
+        eg.user(userId: 2, fullName: 'test user1'),
+        eg.user(userId: 3, fullName: 'test user2'),
+      ];
+
+      await setupPage(tester,
+        users: users,
+        pageUserId: 1,
+        customProfileFields: [mkCustomProfileField(0, CustomProfileFieldType.user)],
+      );
+
+      final findAvatars = find.byType(Avatar);
+      check(findAvatars.evaluate()).length.equals(2 + 1);
```

Comment:
This "2 + 1" feels a bit brittle to me; in particular the "+ 1" is implicitly referring to the fact that the top of the screen also shows an Avatar.

One approach is to make that more explicit by getting the specific user IDs and checking the list:
```suggestion
      final avatars = tester.widgetList<Avatar>(find.byType(Avatar));
      check(avatars.map((w) => w.userId).toList())
        .deepEquals([1, 2, 3]);
```

If some later tweak makes the big image at the top not an `Avatar` widget, the test would still break; but it'd be a lot more evident how it broke and what the intent was, and so how to update the test.